File: Anagrams/areTheseAnagramsHashTables.java
import java.util.*;

class Algorithm {
    public static void main(String[] args) {
        Scanner userInput = new Scanner(System.in);
        System.out.print("Enter string number 1: ");
        String inputString1 = userInput.nextLine();
        System.out.print("Enter string number 2: ");
        String inputString2 = userInput.nextLine();
        System.out.println(areTheseAnagramsHashTables(inputString1, inputString2));
    }
    public static String areTheseAnagramsHashTables(String s1, String s2) {
        HashMap<Character, Integer> frequencyMap1;
        HashMap<Character, Integer> frequencyMap2 = new HashMap<>();
        if (s1.length() != s2.length()) {
            return "These are not anagrams";
        } else {
            frequencyMap1 = new HashMap<>();

            for (Character chA : s1.toCharArray()) {
                if (frequencyMap1.containsKey(chA)) {
                    frequencyMap1.put(chA, frequencyMap1.get(chA) + 1);
                } else {
                    frequencyMap1.put(chA, 1);
                }
            }
            Map<Character, Integer> sortedFrequencyMap1 = new TreeMap<>(frequencyMap1);
            System.out.println(sortedFrequencyMap1);
            for (Character chB : s2.toCharArray()) {
                if (frequencyMap2.containsKey(chB)) {
                    frequencyMap2.put(chB, frequencyMap2.get(chB) + 1);
                } else {
                    frequencyMap2.put(chB, 1);
                }
            }
            Map<Character, Integer> sortedFrequencyMap2 = new TreeMap<>(frequencyMap2);
            System.out.println(sortedFrequencyMap2);
        }
        for (Character chA : s1.toCharArray()) {
            if (!frequencyMap2.containsKey(chA) || !Objects.equals(frequencyMap1.get(chA), frequencyMap2.get(chA))) {
                return "These are not anagrams";
            }
        }
        return "These are anagrams";
    }
}

File: Anagrams/areTheseAnagramsHashTables.swift
@main
public struct theCodeZettaStone {
   public static func main() {
        print("Enter string number 1: ", terminator: "")
        let inputString1: String? = readLine()
        print("Got it. Thanks!")
        print("Enter string number 2: ", terminator: "")
        let inputString2: String? = readLine()
        print(this(inputString1, inputString2))
    }
}

func this(_ s1: String?, _ s2: String?) -> String {
    var frequencyDictionary1: [Character: Int] = [:]
    var frequencyDictionary2: [Character: Int] = [:]
    if s1?.count != s2?.count {
        return "These are not anagrams"
    } else {
        s1?.forEach({ chr in
            if frequencyDictionary1[chr] == nil {
                frequencyDictionary1[chr] = 1
            } else {
                frequencyDictionary1[chr]? += 1
            }
        })
        var counter = 0
        print("[", terminator: "")
        for chr: Character in frequencyDictionary1.keys.sorted() {
            counter += 1
            if counter < frequencyDictionary1.count {
                print("\(chr): \(frequencyDictionary1[chr]!), ", terminator: "")
            } else {
                print("\(chr): \(frequencyDictionary1[chr]!)", terminator: "")
            }
        }
        print("]")
        s2?.forEach({ chr in
            if frequencyDictionary2[chr] == nil {
                frequencyDictionary2[chr] = 1
            } else {
                frequencyDictionary2[chr]? += 1
            }
        })
        counter = 0
        print("[", terminator: "")
        for chr: Character in frequencyDictionary2.keys.sorted() {
            counter += 1
            if counter < frequencyDictionary2.count {
                print("\(chr): \(frequencyDictionary2[chr]!), ", terminator: "")
            } else {
                print("\(chr): \(frequencyDictionary2[chr]!)", terminator: "")
            }
        }
        print("]")
    }
    for key in frequencyDictionary1.keys {
        if !frequencyDictionary2.keys.contains(key) || frequencyDictionary1[key] != frequencyDictionary2[key] {
            return "These are not anagrams"
        }
    }
    return "These are anagrams"
}
